pkg/networkservice/ipam/server.go

```diff
@@ -0,0 +1,160 @@
+// * IP net - assigns ip address out of a pool of prefixes with a CIDR mask of selected prefix
+// The IP addresses assigned are not reassigned until they are released. All subsequent requests return previously
+// assigned IP addresses except than the request type has been changed or existing IP addresses are excluded with new
+// exclude prefixes.
+package ipam
+
+import (
+	"context"
+	"net"
+	"sync"
+
+	"github.com/RoaringBitmap/roaring"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+)
+
+type ipamServer struct {
+	ipPools   []*ipPool
+	connInfos connectionInfoMap
+	prefixes  []*net.IPNet
+	once      sync.Once
+	initErr   error
+}
+
+type connectionInfo struct {
+	ipPool  *ipPool
+	srcAddr string
+	dstAddr string
+}
+
+// NewServer - creates a NetworkServiceServer that requests a kernel interface and populates the netns inode
+func NewServer(prefixes ...*net.IPNet) networkservice.NetworkServiceServer {
+	return &ipamServer{
+		prefixes: prefixes,
+	}
+}
+
+func (s *ipamServer) init() {
+	if len(s.prefixes) == 0 {
+		s.initErr = errors.New("required one or more prefixes")
+		return
+	}
+
+	for _, prefix := range s.prefixes {
+		if prefix == nil {
+			s.initErr = errors.Errorf("prefix must not be nil: %+v", s.prefixes)
+			return
+		}
+		s.ipPools = append(s.ipPools, newIPPool(prefix))
+	}
+}
+
+func (s *ipamServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	s.once.Do(s.init)
+	if s.initErr != nil {
+		return nil, s.initErr
+	}
+
+	conn := request.GetConnection()
+	if conn.GetContext() == nil {
+		conn.Context = &networkservice.ConnectionContext{}
+	}
+	if conn.GetContext().GetIpContext() == nil {
+		conn.GetContext().IpContext = &networkservice.IPContext{}
+	}
+	ipContext := conn.GetContext().GetIpContext()
+
+	exclude, err := exclude(ipContext.GetExcludedPrefixes()...)
+	if err != nil {
+		return nil, err
+	}
+
+	connInfo, ok := s.connInfos.Load(conn.GetId())
+	if ok && shouldUpdate(connInfo, ipContext, exclude) {
+		// we are requested for another { p2p, IP net } case or some of the existing addresses are excluded
+		s.delete(conn.GetId(), connInfo)
+		ok = false
+	}
+	if !ok {
+		var srcAddr, dstAddr string
+		for _, ipPool := range s.ipPools {
+			if ipContext.GetDstIpRequired() {
+				// p2p ipam case
+				if dstAddr, srcAddr, err = ipPool.getP2PAddrs(exclude); err != nil {
+					continue
+				}
+			} else {
+				// IP net ipam case
+				if srcAddr, err = ipPool.getIPNetAddr(exclude); err != nil {
+					continue
+				}
+			}
+			connInfo = &connectionInfo{
+				ipPool:  ipPool,
+				srcAddr: srcAddr,
+				dstAddr: dstAddr,
+			}
+			s.connInfos.Store(conn.GetId(), connInfo)
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Consider moving this logic to a separate private method to reduce nesting complexity.

Reply:
Yep, fixed.